### ISSUE TITLE
Allow indexing for object list

### ISSUE THREAD

**Issue by saikumar-a** [mapstruct/mapstruct#1321]
Allow access to list objects with index like the one below.

```java
public interface MyMapper {
    @Mapping(target = "subjectName", source = "course[0].courseName")
    Target map(Source source);
} 
```

**Comment by filiphr** [comment]
@saikumar-a thanks for raising this issue. I think that someone has already proposed something like this, but I can't seem to find the related issue. In any case, this is currently not supported in MapStruct.

You are free to work on such a feature if you are interested :)

**Comment by gunnarmorling** [comment]
Note you can use expressions as work-around for the time being:

    public interface MyMapper {
        @Mapping(target = "subjectName", expression = "java(source.getCourses().get(0).getCourseName())")
        Target map(Source source);
    }

**Comment by saikumar-a** [comment]
@gunnarmorling thanks for the response. But how can I achieve the target index mapping like the one below. 
```java
public interface MyMapper {
    @Mapping(target = "subject[0].subjectName", source = "courseName")
    Target map(Source source);
} 
```

**Comment by filiphr** [comment]
@saikumar-a it is not possible to write an expression for writing into the target field. In order to achieve what you are looking for you should use one of the lifecycle methods `@BeforeMapping` / `@AfterMapping`.

It will look like:

```java
public interface MyMapper {
    @Mapping(target = "subject[0].subjectName", source = "courseName")
    Target map(Source source);

    @AfterMapping
    default public void after(@MappingTarget Target target, Source source) {
        target.getSubject().get(0).setSubjectName(source.getCourseName());
    }
} 
```

**Comment by chris922** [comment]
@filiphr I just came across this issue because I've got the requirement. Anything that should be considered implementing it?

I would assume that we should implement it like described, access to List entries (or in general Iterables) via square brackets and the index.

@sjaakd : Should it somehow be aligned with your work on the Map support from PR #1744 ?

**Comment by filiphr** [comment]
@chris922 thanks for looking into this.

Not sure if you need the `source` only, or also the `target`. I think that for `target` will be a bit more complex since you can't put elements on a fix location always.

What I will say is that we would need to support this only for `List` and not for any collection type. We should also make sure that the invocation does not throw `IndexOutOfBoundsException`, so we would probably need to check for size of the list when accessing the element.

**Comment by chris922** [comment]
I would propose to start with `source` and move support for `target` to a separate issue if it has a higher complexity. Having this for `source` in place should already be great.

> What I will say is that we would need to support this only for List and not for any collection type.

That makes sense. `List` provides a suitable `get` method, for the other kind of collections or iterables something more complicated would be required.. 

> We should also make sure that the invocation does not throw IndexOutOfBoundsException, so we would probably need to check for size of the list when accessing the element.

And what should we throw instead? I think it makes sense to inform the user that something can't be mapped. Unfortunately we have to do this at runtime, because we don't know the size of the list at compile time

**Comment by filiphr** [comment]
>And what should we throw instead? I think it makes sense to inform the user that something can't be mapped. Unfortunately we have to do this at runtime, because we don't know the size of the list at compile time

I don't think we should throw anything. If the user requested `course[1].name` and the size is not `>=2` then use `null` for the name.

**Comment by chris922** [comment]
I could imagine that this will lead to confusing bugs when something is `null` just because someone misconfigured the mapping.. I personally would prefer an exception over not expected `null`s in some beans and NPE on totally different places.. what about an extra configuration? .. but we already have so many configurations / policies / strategies / ... 😅

**Comment by filiphr** [comment]
I don't think it will lead to confusing bugs if we explain it in the documentation. In the end we are a mapping library which generates Java code. If something is not clear they can always see the generated code. I personally wouldn't go for extra configuration. I don't see this different compared to the other nested source and something being `null` i.e. missing in between

**Comment by K-e-l-v-i-n-12** [comment]
We need access to list elements by index frequently. We have been following the issue with interest for the last few years. Unfortunately, there has been no movement for many years. Could the feature request be resumed?

Many thanks in advance :)
